### PR TITLE
Add support for adding testRunListeners

### PR DESCRIPTION
**Pull request by x2on**
This PR adds support for adding custom `ITestRunListener`.

We need this for our https://github.com/x2on/gradle-spoon-plugin to integrate it in TeamCity and get notifications about the running tests.

### CODE REVIEW

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonRunner.java [hunk 1]
```diff
@@ -322,6 +329,11 @@ public Builder setMethodName(String methodName) {
       return this;
     }
 
+    public Builder setTestRunListeners(List<ITestRunListener> testRunListeners) {
```
Comment (JakeWharton) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonRunner.java]: This should take a single `ITestRunListener` and add it to the list (which is initialized to an empty `ArrayList`).

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java [hunk 2]
```diff
@@ -74,11 +76,12 @@
    * @param className Test class name to run or {@code null} to run all tests.
    * @param methodName Test method name to run or {@code null} to run all tests.  Must also pass
    *        {@code className}.
+   * @param testRunListeners Additional TestRunListener or {@code null}.
```
Comment (JakeWharton) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: This should never be `null`. `null` is death. Empty lists are much better.

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java [hunk 3]
```diff
@@ -198,10 +202,11 @@ public DeviceResult run(AndroidDebugBridge adb) {
...
         runner.setTestSize(testSize);
       }
-      runner.run(
-          new SpoonTestRunListener(result, debug, testIdentifierAdapter),
-          new XmlTestRunListener(junitReport)
-      );
+      List<ITestRunListener> listeners = new ArrayList<ITestRunListener>();
+      listeners.add(new SpoonTestRunListener(result, debug, testIdentifierAdapter));
+      listeners.add(new XmlTestRunListener(junitReport));
+      listeners.addAll(testRunListeners);
```
Comment (JakeWharton) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: Plus this would have NPE'd if `testRunListeners` was actually `null`.

**Comment by JakeWharton** [comment]
I'm fine adding this, just would like the one API tweak to be made.

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java (hunk 3 above)
Comment (x2on) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: `ArrayList` is safe for adding `null`.

Comment (JakeWharton) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: But `addAll` is not.

Comment (x2on) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java]: Ah yeah thats right - thanks for that hint.

**Comment by x2on** [comment]
@JakeWharton changes are made - fine with this?

File: spoon-runner/src/main/java/com/squareup/spoon/SpoonRunner.java [hunk 4]
```diff
@@ -215,6 +222,7 @@ private SpoonDeviceRunner getTestRunner(String serial, SpoonInstrumentationInfo
     private IRemoteAndroidTestRunner.TestSize testSize;
     private int adbTimeout;
     private boolean failIfNoDeviceConnected;
+    private List<ITestRunListener> testRunListeners;
```
Comment (edenman) [inline spoon-runner/src/main/java/com/squareup/spoon/SpoonRunner.java]: You can just initialize this to `new ArrayList<>` and then you don't need the null checks all over the place.

**Comment by x2on** [comment]
@edenman changes are done.

**Comment by edenman** [comment]
Thanks, @x2on!

**Pull request merged**